qutebrowser/browser/bookmarks.py

```diff
@@ -0,0 +1,140 @@
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Manager for bookmarks.
+
+Note we violate our general QUrl rule by storing url strings in the bookmarks
+OrderedDict. This is because we read them from a file at start and write them
+to a file on shutdown, so it makes sense to keep them as strings here.
+"""
+
+import os
+import os.path
+import collections
+
+from PyQt5.QtCore import pyqtSignal, QUrl, QObject
+
+from qutebrowser.utils import message, urlutils, standarddir, objreg
+from qutebrowser.commands import cmdexc
+from qutebrowser.misc import lineparser
+
+
+class BookmarkManager(QObject):
+
+    """Manager for bookmarks.
+
+    Attributes:
+        bookmarks: An OrderedDict of all bookmarks.
+        _lineparser: The LineParser used for the bookmarks, or None
+                     (when qutebrowser is started with -c '').
+
+    Signals:
+        changed: Emitted when anything changed.
+        added: Emitted when a new bookmark was added.
+               arg 0: The title of the bookmark.
+               arg 1: The URL of the bookmark, as string.
+        removed: Emitted when an existing bookmark was removed.
+                 arg 0: The title of the bookmark.
+    """
+
+    changed = pyqtSignal()
+    added = pyqtSignal(str, str)
+    removed = pyqtSignal(str)
+
+    def __init__(self, parent=None):
+        """Initialize and read bookmarks."""
+        super().__init__(parent)
+
+        self.bookmarks = collections.OrderedDict()
+
+        if standarddir.config() is None:
+            self._lineparser = None
+        else:
+            bookmarks_directory = os.path.join(standarddir.config(),
+                                               'bookmarks')
+            if not os.path.isdir(bookmarks_directory):
+                os.makedirs(bookmarks_directory)
+            self._lineparser = lineparser.LineParser(
+                standarddir.config(), 'bookmarks/urls', parent=self)
+            for line in self._lineparser:
+                if not line.strip():
+                    # Ignore empty or whitespace-only lines.
+                    continue
+                try:
+                    url, title = line.split(maxsplit=1)
+                except ValueError:
+                    message.error(0, "Invalid bookmark '{}'".format(line))
+                else:
+                    self.bookmarks[url] = title
+            filename = os.path.join(standarddir.config(), 'bookmarks/urls')
+            objreg.get('save-manager').add_saveable(
+                'bookmark-manager', self.save, self.changed,
+                filename=filename)
+
+    def save(self):
+        """Save the bookmarks to disk."""
+        if self._lineparser is not None:
+            self._lineparser.data = [' '.join(tpl)
+                                     for tpl in self.bookmarks.items()]
+            self._lineparser.save()
+
+    def bookmark_add(self, win_id, url, title):
```

Comment:
This should probably just be called `add` - the `bookmark_` prefix is not needed (it's there for `quickmark_add` because that's registered as a command).